contributors/design-proposals/api-machinery/admission-webhook-GA-proposal.md

```diff
@@ -0,0 +1,488 @@
+# Graduate Admission Webhooks to GA
+
+Authors: @mbohlool
+
+Reviewers: @liggit @deads2k @sttts @caesarxuchao
+
+Sig: api-machinery
+
+## Summary
+
+Admission webhooks are a way to extend kubernetes by putting hooks over object creation/modification. Admission webhooks can mutate or validate the object. This feature has been Beta since Kubernetes 1.9. This document outline required steps to graduate it to GA.
+
+## New Features
+
+Based on users feedback, these are the changes that are proposed to graduate Admission Webhooks to GA.
+
+### Object selector
+
+Currently Admission Webhook supports namespace selectors, but that may not be enough for some cases that admission webhook need to be skipped on some objects. For example if the Admission Webhook is running inside cluster and its rules includes wildcards which match required API objects for its own execution. An object selector would be useful exclude those objects. Also in case of an optional webhooks, an object selector gives the end user to include or exclude an object without having access to admission webhook configuration which is normally restricted to cluster admins.
+Note that namespace objects must match both object selector (if specified) and namespace selector to be sent to the Webhook except for cluster-scoped resources which namespace selector will not applied to. The propsoe change is to add an ObjectSelector to the webhook API both in v1 and v1beta1.
+
+```golang
+type Webhook struct {
+    ...
+     // ObjectSelector decides whether to run the webhook on an object based
+     // on whether the object.metadata.labels matches the selector. An object is
```

Comment:
describe which object the selector is run against for each mutating operation:
- resources:
  - create: incoming object
  - update: incoming object?
  - delete: existing object? (delete admission doesn't have the old object available to it today)
- subresources that accept an object with labels (e.g. pods/status):
  - create: incoming object?
  - update: incoming object?
  - delete: old object? (delete admission doesn't have the old object available to it today)
- subresources that don't accept an object with labels (e.g. pods/proxy):
  - create: existing parent object? (I don't think admission has this info today)
  - update: existing parent object? (I don't think admission has this info today)
  - delete: existing parent object? (I don't think admission has this info today)

Reply:
for delete, the options I see are:
1. plumb existing object to delete admission (which is nice for admission authors and sort of thorny for apiserver authors... see https://github.com/kubernetes/kubernetes/pull/27193#issuecomment-225270110)
2. not route delete calls to objectSelector-filtered admission (which I would find super-unexpected as a webhook author)
3. not allow objectSelector on admission plugins asking to be sent DELETE events

selection on subresources gets weird... not sure what to do there

Reply:
We discussed Object Selectors for specific usecases. This shows it is more complicated than we thought. What about not supporting Object Selector for sub-resources and Delete operation? (with validation on them of course). Would it be too much restrictive?

Reply:
outlining the primary use cases for this would help determine if disallowing objectSelector combined with DELETE operations or subresources (side-stepping the issue) would be limiting